DOCSP-34027 Change Streams (#223)
mongodb/docs-csharp: source/fundamentals/crud/read-operations/change-streams.txt

.. _csharp-change-streams:

====================
Monitor Data Changes
====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: watch, code example

Overview
--------

In this guide, you can learn how to use a **change stream** to monitor real-time

> changes to your database. A change stream is a {+mdb-server+} feature that

Reviewer: S: Since the next sentence clarifies that you can monitor collection, database, or deployment changes, I think it's confusing to just refer to database monitoring here. Maybe "data" would work better:
```suggestion
changes to your data. A change stream is a {+mdb-server+} feature that
```

classes:

- ``MongoClient``: To monitor all changes in the MongoDB deployment
- ``Database``: To monitor changes in all collections in the database
- ``Collection``: To monitor changes in the collection

The following example opens a change stream on the ``restaurants`` collection
and outputs the operation type of changes as they occur. Select the
:guilabel:`Asynchronous` or :guilabel:`Synchronous` tab to see the corresponding
code.

.. tabs::

   .. tab:: Asynchronous
      :tabid: change-stream-async

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-open-change-stream-async
         :end-before: end-open-change-stream-async
         :language: csharp

   .. tab:: Synchronous
      :tabid: change-stream-sync

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-open-change-stream
         :end-before: end-open-change-stream
         :language: csharp

To begin watching for changes, run the application. Then, in a separate
application or shell, modify the ``restaurants`` collection. The following

> example updates a document with a ``name`` field value of ``Blarney Castle``:

Reviewer: S: to avoid "with"; also, I'd add quotes around the string value:

```suggestion
example updates a document that has a ``name`` field value of ``"Blarney Castle"``:
```

.. _csharp-change-stream-update:

.. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
   :start-after: start-modify-document
   :end-before: end-modify-document
   :language: csharp

> When you update the collection, the change stream application prints the type of change

Reviewer: S: Since this section is focused on the change stream rather than updating documents, it might be more helpful to include the change stream output produced by modifying the document instead of an example of how to modify a document. [This section](https://www.mongodb.com/docs/drivers/go/current/fundamentals/crud/read-operations/changestream/#example) on the Go page has just the output, for example

Author: Updated the code a bit to have a more relevant output and added the output in place of the update code

as it occurs.

Modify the Change Stream Output
-------------------------------

You can pass the ``pipeline`` parameter to the ``Watch()`` and ``WatchAsync()``
methods to modify the change stream output. This parameter allows you to watch
for only specified change events. Create the pipeline by using the

> ``EmptyPipelineDefinition`` class, and appending the relevant aggregation stage methods.

Reviewer: S: I don't think you need a comma here
```suggestion
``EmptyPipelineDefinition`` class and appending the relevant aggregation stage methods.
```

You can specify the following aggregation stages in the ``pipeline`` parameter:

- ``$addFields``
- ``$match``
- ``$project``
- ``$replaceRoot``
- ``$replaceWith``
- ``$redact``
- ``$set``
- ``$unset``

To learn how to build an aggregation pipeline by using the

> ``PipelineDefinitionBuilder`` class, see :ref:`csharp-builders-aggregation`.

Reviewer: S: specify where this link is going
```suggestion
``PipelineDefinitionBuilder`` class, see :ref:`csharp-builders-aggregation` in the Operations with Builders guide.
```

.. tabs::

   .. tab:: Asynchronous
      :tabid: change-stream-async

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-change-stream-pipeline-async
         :end-before: end-change-stream-pipeline-async
         :language: csharp

   .. tab:: Synchronous
      :tabid: change-stream-sync

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-change-stream-pipeline
         :end-before: end-change-stream-pipeline
         :language: csharp

To learn more about modifying your change stream output, see the
:manual:`Modify Change Stream Output
</changeStreams/#modify-change-stream-output>` section in the {+mdb-server+}
manual.

Modify ``Watch()`` Behavior
---------------------------

The ``Watch()`` and ``WatchAsync()`` methods accept optional parameters, which represent
options you can use to configure the operation. If you don't specify any
options, the driver does not customize the operation.

The following table describes the options you can set to customize the behavior
of ``Watch()`` and ``WatchAsync()``:

.. list-table::
   :widths: 30 70
   :header-rows: 1

>    * - Property

Reviewer: S: since you refer to them as options in the copy, I'd update the table header
```suggestion
   * - Option
```

         operation specified in the resume token.
       | Each change stream event document includes a resume token as the ``_id``
         field. Pass the entire ``_id`` field of the change event document that
         represents the operation you want to resume after.
       | ``ResumeAfter`` is mutually exclusive with ``StartAfter`` and ``StartAtOperationTime``.

   * - ``StartAfter``
     - | Directs ``Watch()`` or ``WatchAsync()`` to start a new change stream after the
         operation specified in the resume token. Allows notifications to
         resume after an invalidate event.
       | Each change stream event document includes a resume token as the ``_id``
         field. Pass the entire ``_id`` field of the change event document that
         represents the operation you want to resume after.
       | ``StartAfter`` is mutually exclusive with ``ResumeAfter`` and ``StartAtOperationTime``.
       
   * - ``StartAtOperationTime``
     - | Directs ``Watch()`` or ``WatchAsync()`` to return only events that occur after the
         specified timestamp.
       | ``StartAtOperationTime`` is mutually exclusive with ``ResumeAfter`` and ``StartAfter``.

   * - ``MaxAwaitTime``

>      - | The maximum amount of time, in milliseconds, the server waits for new

Reviewer: S: since the first few descriptions start with a verb, it would be good to keep that consistent throughout the descriptions

```suggestion
     - | Specifies the maximum amount of time, in milliseconds, the server waits for new
```

Reviewer: Applies to all

- ``ChangeStreamFullDocumentOption.WhenAvailable``: The change event includes a
  post-image of the modified document for change events only if the post-image
  is available.
- ``ChangeStreamFullDocumentOption.Required``: The change event includes a
  post-image of the modified document for change events. If the post-image is
  not available, the driver raises an error.

The following example opens a change stream on a collection and includes the post-image
of updated documents by specifying the ``FullDocument`` option. Select the
:guilabel:`Asynchronous` or :guilabel:`Synchronous` tab to see the corresponding
code.

.. tabs::

   .. tab:: Asynchronous
      :tabid: change-stream-async

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-change-stream-post-image-async
         :end-before: end-change-stream-post-image-async
         :language: csharp

   .. tab:: Synchronous
      :tabid: change-stream-sync

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-change-stream-post-image
         :end-before: end-change-stream-post-image
         :language: csharp

> 

Reviewer: S: I think it would be helpful to see what the post-image looks like in the output

Author: Added!

classes:

- ``MongoClient``: To monitor all changes in the MongoDB deployment
- ``Database``: To monitor changes in all collections in the database
- ``Collection``: To monitor changes in the collection

The following example opens a change stream on the ``restaurants`` collection
and outputs the changes as they occur. Select the
:guilabel:`Asynchronous` or :guilabel:`Synchronous` tab to see the corresponding
code.

.. tabs::

   .. tab:: Asynchronous
      :tabid: change-stream-async

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-open-change-stream-async
         :end-before: end-open-change-stream-async
         :language: csharp

   .. tab:: Synchronous
      :tabid: change-stream-sync

      .. literalinclude:: /includes/code-examples/change-streams/change-streams.cs
         :start-after: start-open-change-stream
         :end-before: end-open-change-stream
         :language: csharp

To begin watching for changes, run the application. Then, in a separate
application or shell, modify the ``restaurants`` collection. Updating a document

> with the ``"name"`` value of ``"Blarney Castle"`` results in the following

Reviewer: S: to avoid "with" (applies to all)
```suggestion
that has a ``"name"`` value of ``"Blarney Castle"`` results in the following
```

change stream output:

> .. code-block:: sh

Reviewer: S: I'd add copyable:: false to the outputs (applies to all)

```suggestion
.. code-block:: sh
   :copyable: false
```

       
   * - ``StartAtOperationTime``
     - | Directs ``Watch()`` or ``WatchAsync()`` to return only events that occur after the
         specified timestamp.
       | ``StartAtOperationTime`` is mutually exclusive with ``ResumeAfter`` and ``StartAfter``.

   * - ``MaxAwaitTime``
     - | Specifies the maximum amount of time, in milliseconds, the server waits for new
         data changes to report to the change stream cursor before returning an
         empty batch. Defaults to 1000 milliseconds.

   * - ``ShowExpandedEvents``
     - | Starting in {+mdb-server+} v6.0, change streams support change notifications
         for Data Definition Language (DDL) events, such as the ``createIndexes`` and ``dropIndexes`` events. To
         include expanded events in a change stream, create the change stream
         cursor and set this parameter to ``True``.

   * - ``BatchSize``
     - | Specifies the maximum number of change events to return in each batch of the
         response from the MongoDB cluster.

   * - ``Collation``
     - | Specifies the collation to use for the change stream cursor.

   * - ``Comment``

>      - | Specifies a comment to attach to the operation.

Reviewer: S: just to make this more concise:

```suggestion
     - | Attaches a comment to the operation.
```
